feat: Cast elements to correct type when creating vectors

aleksanderkatan opened the pull request.

aleksanderkatan commented:
All those checks seem to have affected the performance a bit

<img width="1486" alt="image" src="https://github.com/user-attachments/assets/338207ce-946e-413a-8a35-0366960c4388" />

mhawryluk reviewed (approved):
Well done!! 🚀

reczkok reviewed (commented):
I'm really on the fence about this — I get that the consistency between JS and WGSL is nice, but is it really worth the performance hit (which, as we've established, is kind of important)?

reczkok commented:
> I'm really on the fence about this — I get that the consistency between JS and WGSL is nice, but is it really worth the performance hit (which, as we've established, is kind of important)?

Maybe we could leave this in and figure out a way to provide users with a very fast mass vector constructor

aleksanderkatan commented:
I think we could modify the schema casts for scalar values, so that `f32` and `f16` are identity, `i32` is `Math.floor(n)`, `u32` is `Math.max(0, Math.floor(n))`, and `bool` is left as is (all plus the undefined check). This would probably increase the performance and leave the most important part of the cast in place. What do you think?

reczkok commented:
I think we should keep the casting logic to stay compliant with WGSL's behavior. I’m actually curious how much of the performance hit comes from the casting itself versus the overhead of the function calls.

aleksanderkatan commented on packages/typegpu/src/data/numeric.ts:
```diff
@@ -46,7 +76,9 @@ const u32Cast = createDualImpl(
   },
   // GPU implementation
-  (v) => snip(`u32(${v.value})`, u32),
+  (v) => {
+    return snip(`u32(${v?.value ?? ''})`, u32);
+  },
```
mental missclick

aleksanderkatan commented:
@reczkok seems like most of the overhead is in Math functions and array writes.

Before any changes:
<img width="1487" alt="image" src="https://github.com/user-attachments/assets/bbf61f39-67e3-442e-b025-6407196d3947" />

When calling schema constructors, but all they do is return the value passed in:
<img width="1487" alt="image" src="https://github.com/user-attachments/assets/d1022ad4-6db6-4855-953f-f1d90be07602" />

When calling schema constructors, and casting all values properly:
<img width="1487" alt="image" src="https://github.com/user-attachments/assets/4780a07b-ba03-45b9-abc5-5cdf1db60bb7" />

aleksanderkatan commented:
The benchmark results are now wildly inconsistent. For one test, I get 30 ops/s five times in a row, then I get 11 ops/s five times in a row...

I added the coercion to the vector element assignment (thgough `[0]` and through `.x`). I also removed the float value casts, and I simplified a bit the integer casts.

aleksanderkatan commented on packages/typegpu/src/data/vectorImpl.ts:
```diff
@@ -1,11 +1,14 @@
 export abstract class VecBase<S> extends Array implements SelfResolvable {
   public readonly [$internal] = true;
   abstract get kind(): VecKind;
+  abstract getElementSchema(): (v?: S) => S;
```
It was meant to be a getter but I don't think this is possible, unless I am missing something
<img width="1253" alt="image" src="https://github.com/user-attachments/assets/4aff2c3d-65e3-42af-9f8c-4122fdc677a4" />

iwoplaza replied:
Aah, I see, I did not know about this limitation. It's good then!

iwoplaza reviewed (approved):
This is a great step forward 👏
